buildSrc/subprojects/build/src/main/groovy/org/gradle/build/BuildReceipt.groovy

```diff
@@ -32,6 +32,7 @@ class BuildReceipt extends DefaultTask {
     private static final SimpleDateFormat TIMESTAMP_FORMAT = new java.text.SimpleDateFormat('yyyyMMddHHmmssZ')
     private static final SimpleDateFormat ISO_TIMESTAMP_FORMAT = new java.text.SimpleDateFormat('yyyy-MM-dd HH:mm:ss z')
     static {
+        TIMESTAMP_FORMAT.setTimeZone(TimeZone.getTimeZone("UTC"));
```

Comment:
@wolfs As you predicted, I was bitten by the Gradle version inconsistence - the generated build receipt is `20190108053720+0100` but when it's read by the downstream tasks it becomes `20190108043720+0100`, thus breaking cacheability. Can we use ISO here?

Reply:
What do you mean? IIUC, the timestamp is read from the `buildReceipt.properties` file by all downstream builds of `CompileAll`, so it should be consistent. Is this not the case? Why would `ISO` help?

Reply:
OK... I thought you knew this issue... sorry for not making myself clear, let me clarify:

When the timestamp is generated in `compileAll` build, it's not UTC but UTC+1 - something like `20190108053720+0100`: https://builds.gradle.org/viewLog.html?buildId=18895776&buildTypeId=Gradle_Check_CompileAll&tab=artifacts . In `compileAll` build, `rootProject.version=5.2-20190108053720+0100`.

However, in downstream builds, when `buildReceipt.properties` is read, it's parsed as UTC: https://github.com/gradle/gradle/blob/1e4bb4edaaa74cde83ef0063e0b40f34dda1ae2f/gradle/versioning.gradle#L26 . Therefore, in the downstream builds, `rootProject.version=5.2-20190108043720+0000`.

This inconsistence break the cacheability here: https://github.com/gradle/gradle/blob/46ed7414de1c8dd0c5e6027a6cf351d882ee44e0/subprojects/docs/docs.gradle#L375 when I was trying to put `distDocs` task into `compileAllBuild`.

Reply:
Can you trace the writing and reading of the timestamp and maybe build up a unit test? A quick look showed that the timestamp is written with timezone UTC and read with timezone UTC again. So the result should be the same right? I am probably missing something in between.

Reply:
So, adding the line we are commenting on fixes the issue?
```
TIMESTAMP_FORMAT.setTimeZone(TimeZone.getTimeZone("UTC"));
```

If yes, then that is perfectly fine to use!

Reply:
Yes. I figure out what's going on here:

When `compileAllBuild` task runs, it [formats current time](https://github.com/gradle/gradle/blob/1e4bb4edaaa74cde83ef0063e0b40f34dda1ae2f/gradle/versioning.gradle#L40) as UTC and uses this timestamp string as internal version `5.2-20190108043720+0000`, but when it writes to receipt file, it formats it in UTC+1 timezone as `20190108053720+0100` so all downstream builds reading the receipt get `5.2-20190108043720+0000`. 

We need to either change [this formatter](https://github.com/gradle/gradle/blob/1e4bb4edaaa74cde83ef0063e0b40f34dda1ae2f/gradle/versioning.gradle#L27) to machine timezone or change BuildReceipt formatter to UTC. Let's prefer the latter way.